renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -0,0 +1,607 @@
+          (VkImage)m_pDriver->GetResourceManager()->GetCurrentResource(depthImage);
+      depthCopyParams.srcImageLayout = VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;
+      depthCopyParams.srcImageFormat = imginfo.format;
+      m_pDriver->GetDebugManager()->PixelHistoryCopyPixel(cmd, depthCopyParams, offset + 16);
+    }
+  }
+
+  void PreDraw(uint32_t eid, VkCommandBuffer cmd)
+  {
+    auto it = m_Events.find(eid);
+    if(it == m_Events.end())
+      return;
+    EventUsage event = it->second;
+    m_PrevState = m_pDriver->GetRenderState();
+
+    m_pDriver->GetRenderState().EndRenderPass(cmd);
+    // Get pre-modification values
+    uint32_t storeOffset = m_EventIndex * sizeof(EventInfo);
+    CopyPixel(eid, cmd, storeOffset);
+
+    // TODO: add all logic for draw calls.
+
+    m_pDriver->GetRenderState() = m_PrevState;
+    if(m_PrevState.graphics.pipeline != ResourceId())
+      m_pDriver->GetRenderState().BeginRenderPassAndApplyState(cmd, VulkanRenderState::BindGraphics);
+  }
+
+  bool PostDraw(uint32_t eid, VkCommandBuffer cmd)
+  {
+    if(m_Events.find(eid) == m_Events.end())
+      return false;
+
+    m_pDriver->GetRenderState().EndRenderPass(cmd);
+
+    uint32_t storeOffset = m_EventIndex * sizeof(EventInfo);
+    CopyPixel(eid, cmd, storeOffset + offsetof(struct EventInfo, postmod));
+
+    m_pDriver->GetRenderState().BeginRenderPassAndApplyState(cmd, VulkanRenderState::BindGraphics);
+
+    // Get post-modification values
+    m_EventIndex++;
+    return false;
+  }
+
+  void PostRedraw(uint32_t eid, VkCommandBuffer cmd)
+  {
+    // nothing to do
+  }
+
+  void PreDispatch(uint32_t eid, VkCommandBuffer cmd)
+  {
+    if(m_Events.find(eid) == m_Events.end())
+      return;
+
+    uint32_t storeOffset = m_EventIndex * sizeof(EventInfo);
+    CopyPixel(eid, cmd, storeOffset);
+  }
+  bool PostDispatch(uint32_t eid, VkCommandBuffer cmd)
+  {
+    if(m_Events.find(eid) == m_Events.end())
+      return false;
+
+    uint32_t storeOffset = m_EventIndex * sizeof(EventInfo);
+    CopyPixel(eid, cmd, storeOffset + offsetof(struct EventInfo, postmod));
+    m_EventIndex++;
+    return false;
+  }
+  void PostRedispatch(uint32_t eid, VkCommandBuffer cmd) {}
+  void PreMisc(uint32_t eid, DrawFlags flags, VkCommandBuffer cmd)
+  {
+    if(m_Events.find(eid) == m_Events.end())
+      return;
+
+    uint32_t storeOffset = m_EventIndex * sizeof(EventInfo);
+    CopyPixel(eid, cmd, storeOffset);
+  }
+  bool PostMisc(uint32_t eid, DrawFlags flags, VkCommandBuffer cmd)
+  {
+    if(m_Events.find(eid) == m_Events.end())
+      return false;
+
+    uint32_t storeOffset = m_EventIndex * sizeof(EventInfo);
+    CopyPixel(eid, cmd, storeOffset + offsetof(struct EventInfo, postmod));
+    m_EventIndex++;
+    return false;
+  }
+
+  void PostRemisc(uint32_t eid, DrawFlags flags, VkCommandBuffer cmd) {}
+  void PreEndCommandBuffer(VkCommandBuffer cmd) {}
+  void AliasEvent(uint32_t primary, uint32_t alias)
+  {
+    // don't care
```

Comment:
We should care about aliases, and it poses a particular problem for pixel history - the aliased call could easily have different pre/post modification values but we have no infrastructure to record a different copy. I'm not sure how best to solve this - pixel history is likely going to need some ability to record each submitted instance of a command buffer separately with a different copy, or else to do an indirected copy via vkCmdDispatch() that can then be re-based between each submission.

This also ties slightly into the above concern about secondary command buffers.

Reply:
Thanks for the feedback ! I added a TODO for now.